hotungkhanh/jetedge
# KAN-80 feat(timetables): implement backend logic for timetable update

### Ways to Test
#### With frontend
- Run both frontend and backend
- Generate timetable with frontend UI 
- Make modifications in drag and drop
- Click save
- Go to backend endpoint `/timetabling/view` and verify that the changes are saved in the database
- Rinse and repeat do it a few times. Please stress test

#### Single units (the manual way)
Test a successful update:
- [Video example](https://www.loom.com/share/20f0dabdff574ad8a5b63d8a95cae951?sid=6ac90154-b366-4988-9e10-a9d5326cf8d4)
- Generate a timetable (via frontend or backend) to populate the database
- Find a unit that you want to modify (go to backend endpoint `/timetabling/view` and choose)
- Open PostmanAPI and set your request as a `PUT` request
- Construct a JSON according to this format at the bottom, give it an appropriate unitId and new time/location
- Send it, ensure you get a `200 OK` response
- Refresh `/timetabling/view` and check that the unit has been updated

Test unsuccessful update:
- Similar to above, but in your JSON include an invalid buildingId-roomCode pair
- Send it, ensure you get a `404 NOT FOUND` response

#### Construct your JSON
- The attributes that backend will care about are: 
- Unit: `unitId, dayOfWeek, startTime`
- Room: `buildingId, roomCode`
- Go wild with the other attributes the backend doesn't look at them

```
{
    "campus": "",
    "name": "",
    "course": "",
    "duration": -1,
    "students": [],
    "wantsLab": false,
    "unitId": 0,
    "room": {
        "campus": "Geelong",
        "buildingId": "building 2",
        "roomCode": "16",
        "capacity": -1,
        "lab": false
    },
    "dayOfWeek": "TUESDAY",
    "startTime": "14:00:00",
    "end": "15:00:00"
}
```
